// angular-instantsearch/esm5/numeric-selector/numeric-selector.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, Input, Inject, forwardRef } from '@angular/core';
import { connectNumericSelector } from 'instantsearch.js/es/connectors';
import { BaseWidget } from '../base-widget';
import { NgAisInstantSearch } from '../instantsearch/instantsearch';
import { noop } from '../utils';
var NgAisNumericSelector = /** @class */ (function (_super) {
    tslib_1.__extends(NgAisNumericSelector, _super);
    function NgAisNumericSelector(instantSearchParent) {
        var _this = _super.call(this, 'NumericSelector') || this;
        _this.instantSearchParent = instantSearchParent;
        _this.operator = '=';
        _this.state = {
            currentRefinement: null,
            options: [],
            refine: noop,
        };
        return _this;
    }
    /**
     * @return {?}
     */
    NgAisNumericSelector.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.createWidget(connectNumericSelector, {
            attributeName: this.attribute,
            operator: this.operator,
            options: this.items,
        });
        _super.prototype.ngOnInit.call(this);
    };
    NgAisNumericSelector.decorators = [
        { type: Component, args: [{
                    selector: 'ais-numeric-selector',
                    template: "\n    <div [class]=\"cx('')\">\n      <select\n        [class]=\"cx('select')\"\n        (change)=\"state.refine($event.target.value)\"\n      >\n        <option\n          [class]=\"cx('option')\"\n          *ngFor=\"let item of state.options\"\n          [value]=\"item.value\"\n          [selected]=\"item.value === state.currentRefinement\"\n        >\n          {{item.label}}\n        </option>\n      </select>\n    </div>\n  ",
                },] },
    ];
    /** @nocollapse */
    NgAisNumericSelector.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [forwardRef(function () { return NgAisInstantSearch; }),] }] }
    ]; };
    NgAisNumericSelector.propDecorators = {
        attribute: [{ type: Input }],
        operator: [{ type: Input }],
        items: [{ type: Input }]
    };
    return NgAisNumericSelector;
}(BaseWidget));
export { NgAisNumericSelector };
if (false) {
    /** @type {?} */
    NgAisNumericSelector.prototype.attribute;
    /** @type {?} */
    NgAisNumericSelector.prototype.operator;
    /** @type {?} */
    NgAisNumericSelector.prototype.items;
    /** @type {?} */
    NgAisNumericSelector.prototype.state;
    /** @type {?} */
    NgAisNumericSelector.prototype.instantSearchParent;
}
