// videogular2/src/controls/vg-mute/vg-mute.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var vg_mute_1 = require("./vg-mute");
var vg_api_1 = require("../../core/services/vg-api");
describe('Mute Button', function () {
    var mute;
    var ref;
    var api;
    beforeEach(function () {
        ref = {
            nativeElement: {
                getAttribute: function (name) {
                    return name;
                }
            }
        };
        api = new vg_api_1.VgAPI();
        api.medias = {
            main: {
                id: 'main',
                volume: 1
            },
            secondary: {
                id: 'secondary',
                volume: 0.5
            }
        };
        mute = new vg_mute_1.VgMute(ref, api);
    });
    it('Should get media by id on init', function () {
        spyOn(api, 'getMediaById').and.callFake(function () {
            return {
                volume: 1
            };
        });
        mute.vgFor = 'test';
        mute.onPlayerReady();
        expect(api.getMediaById).toHaveBeenCalledWith('test');
        expect(mute.currentVolume).toBe(1);
    });
    it('Should get volume for one media file', function () {
        api.medias = {
            main: {
                volume: 1
            }
        };
        mute.target = api;
        var volume = mute.getVolume();
        expect(volume).toBe(1);
    });
    describe('onClick (single media)', function () {
        it('should mute volume if current volume is different than 0', function () {
            api.medias = {
                main: {
                    volume: 0.75
                }
            };
            mute.target = api;
            mute.onClick();
            expect(mute.currentVolume).toBe(0.75);
            expect(api.volume).toEqual(0);
        });
        it('should unmute volume if current volume is 0', function () {
            api.medias = {
                main: {
                    volume: 0
                }
            };
            mute.target = api;
            mute.currentVolume = 0.75;
            mute.onClick();
            expect(api.volume).toEqual(0.75);
        });
    });
    describe('onClick (multiple medias)', function () {
        it('should mute volume if current volume is different than 0', function () {
            mute.target = api;
            mute.onClick();
            expect(mute.currentVolume).toBe(1);
            expect(api.volume).toEqual(0);
        });
        it('should unmute volume if current volume is 0', function () {
            api.medias = {
                main: {
                    id: 'main',
                    volume: 0
                },
                secondary: {
                    id: 'secondary',
                    volume: 0
                }
            };
            mute.target = api;
            mute.currentVolume = 0.75;
            mute.onClick();
            expect(api.volume).toEqual(0.75);
        });
    });
});
